### ISSUE TITLE
<Datagrid> useListContext(props) returns the wrong resource prop

### ISSUE THREAD

**Issue by wmwart** [marmelab/react-admin#5027]
**What you were expecting:**
I expected that when using `<Datagrid />` inside `<ReferenceManyField />` the translation of the columns would be correct.

**What happened instead:**
Сolumn translation stopped working. This happens because `useListContext(props)` returns the wrong resource prop inside `<Datagrid />`

**Steps to reproduce:**
use the `<Datagrid />` inside `<ReferenceManyField />` as in the documentation

**Related code:**
in ConsumerShow.js
```
...
<ReferenceManyField reference="History" target="item" >
    <ReferenceHistoryDatagrid />
</ReferenceManyField>
...
```

```
export const ReferenceDatagrid = (props) => {
    const classes = useStyles();
    return (
    <Datagrid {...props} rowClick="show" classes={{row:classes.row}}>
        <TextField source="action" noWrap/>
    </Datagrid>
    );
}
```
pay attention to the props that are passed to `<Datagrid />`
![image](https://user-images.githubusercontent.com/28799755/87042258-1830eb80-c1fc-11ea-8a03-1b971f0ba433.png)

and what `useListContext(props)` does inside the `<Datagrid />`
![image](https://user-images.githubusercontent.com/28799755/87042922-fb48e800-c1fc-11ea-9539-0a54f4e19bde.png)
![image](https://user-images.githubusercontent.com/28799755/87043014-216e8800-c1fd-11ea-8e49-a60076b0b8cf.png)

simple example in CodeSandbox not working

**Environment**

* React-admin version: 3.7.0
* Last version that did not exhibit the issue (if applicable): 3.6.2

**Comment by wmwart** [comment]
controversial decision to use `useListContext(props)` inside all the "dumb" components associated with lists (for example `<DataGrid>`). Thus, you tied everything to use within the context. Previously, I could reuse your components anywhere in the application by slipping the correct properties. Now it is impossible.
Why not use `useListContext(props)` outside of these components and transfer properties if necessary?
Will have to roll back to 3.6.x. since the project uses a lot of tables that render data that is in no way related to `ListContext` (

**Comment by fzaninotto** [comment]
Thanks for the bug report. There is indeed a bug in useReferenceManyFieldController, I've just submitted a PR to fix it. 

I'm sorry that you disagree with our context approach. We've made the decision to stop cloning elements and passing props down for 2 reasons:

1. It makes Typescript typing very hard (or useless)
2. It causes performance problems

Using a context (with selectors, which is a future enhancements) solves both these problems, AND makes the building of custom UIs much more natural. 

We've kept the ability to inject props if you haven't wrapped your component with ListContext, so the change should be backwards compatible for now. But we will probably remove that possibility (e.g. to have lower-level components in the Datagrid subscribe to the ListContext directly, without having to pass all the props down) in v4. If you want to use Datagrid outside of react-admin use cases (List, ReferenceManyField, etc), which is not a supported usage, you will have to wrap it inside a ListContext.
